crates/edr_provider/src/requests/serde.rs

```diff
@@ -347,6 +347,63 @@ where
     Ok(result)
 }
 
+/// Value representing timestamp in seconds.
+///
+/// While technically the Yellow Paper allows specifying the timestamp as
+/// any scalar value, using `u64` in practice is more than enough.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Serialize)]
+pub struct Timestamp(pub u64);
+
+impl From<u64> for Timestamp {
+    fn from(value: u64) -> Self {
+        Self(value)
+    }
+}
+
+impl From<Timestamp> for u64 {
+    fn from(value: Timestamp) -> Self {
+        value.0
+    }
+}
+
+impl Timestamp {
+    pub fn as_u64(&self) -> u64 {
+        self.0
+    }
+}
+
+impl<'de> Deserialize<'de> for Timestamp {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserialize_timestamp_u64(deserializer).map(Timestamp)
+    }
+}
+
+fn deserialize_timestamp_u64<'de, D>(deserializer: D) -> Result<u64, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let error_msg = || {
+        serde::de::Error::custom(
+            "Timestamp must be a positive integer or a string not exceeding 2^64.",
+        )
+    };
+
+    match serde_json::Value::deserialize(deserializer)? {
+        serde_json::Value::Number(num) => num.as_u64().ok_or_else(error_msg),
+        serde_json::Value::String(value) => {
+            if let Some(hex_str) = value.strip_prefix("0x") {
+                u64::from_str_radix(hex_str, 16).map_err(|_err| error_msg())
+            } else {
+                value.parse::<u64>().map_err(|_err| error_msg())
+            }
```

Comment:
This deviates from what used to happen. We don't want to allow any type of string here, but one that conforms to the Ethereum format: `^0x([1-9a-f]+[0-9a-f]*|0)$` (also see [this](https://github.com/NomicFoundation/hardhat/blob/ethereumjs/issue4970/packages/hardhat-core/src/internal/core/jsonrpc/types/base-types.ts#L208)).

I recommend using the `U64` type to deserialize the string here, as that already implements this logic. All you'd have to do is to first check the number of bytes in the hex string after `0x` (this part is required). If it exceeds 8 bytes, then return the custom error.

Reply:
I'm also seeing some more inconsistencies, for example we only have `Option<u64>` using default Deserialize in `AccountOverrideOptions` for `eth_call` https://github.com/NomicFoundation/edr/blob/eb00b1533b75134cc453661f22e70d8f31add309/crates/edr_rpc_eth/src/override.rs#L18

but the [official API schema](https://ethereum.github.io/execution-apis/api-documentation/) says it accepts only a hex string of the quantity format you mentioned:

![image](https://github.com/NomicFoundation/edr/assets/3093213/d9faca49-dd13-40bd-a21e-88b151552c59)

Maybe it'd make sense to separate a work item of "go over all of our (de)serialize logic and make sure we're compliant"?

On the other hand per the common networking adage of "be conservative in what you send, be liberal in what you accept" and the fact that we're supposed to be a development platform, I'm not sure if us not being as strict as the spec is necessary the bad thing here.

I'll add it for the agenda for tomorrow's sync; maybe @fvictorio you have some more context or opinions on this.

Reply:
The example you wrote is valid. `0x01` is a single byte equal to 1. There are no leading zeroes.

Reply:
> I'm also seeing some more inconsistencies, for example we only have `Option<u64>` using default Deserialize in `AccountOverrideOptions` for `eth_call`
> 
> https://github.com/NomicFoundation/edr/blob/eb00b1533b75134cc453661f22e70d8f31add309/crates/edr_rpc_eth/src/override.rs#L18
> 
> but the [official API schema](https://ethereum.github.io/execution-apis/api-documentation/) says it accepts only a hex string of the quantity format you mentioned:
> 
> Maybe it'd make sense to separate a work item of "go over all of our (de)serialize logic and make sure we're compliant"?
> 
> On the other hand per the common networking adage of "be conservative in what you send, be liberal in what you accept" and the fact that we're supposed to be a development platform, I'm not sure if us not being as strict as the spec is necessary the bad thing here.
> 
> I'll add it for the agenda for tomorrow's sync; maybe @fvictorio you have some more context or opinions on this.

There is a special deserialize function used in the linked file:

```rust
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        with = "edr_eth::serde::optional_u64"
    )]
    /// Account nonce override.
    pub nonce: Option<u64>,
```
That uses `U64`'s deserialization under the hood.

Reply:
> The example you wrote is valid. 0x01 is a single byte equal to 1. There are no leading zeroes.

But that's not true, right? It:
- neither matches the `^0x([1-9a-f]+[0-9a-f]*|0)$` regexp you pasted and what's in the spec (`0x` followed by either a single `0` or a non-zero amount of `1-f` followed by a possibly zero amount of `0-f`), 
- nor the value seems to be concerned about the bytes per se, specifically `01` in base 10 has leading zeroes in the same sense `01` has as well in base 16. I didn't find a rule that requires us to zero-pad us to whole bytes (and the regex from the spec straight up invalidates it)

> They have additional logic on top of the FromStr implementation.

If I'm not missing something, they seem to handle null strings but other than that, in the `visit_str` they just call `FromStr::from_str` implicitly via `str::parse::<F: FromStr>`, so back to the linked function above and again: `0x01` seems to be converted correctly with the leading zero.

> That uses U64's deserialization under the hood.

Ah, I missed the attributes above :facepalm: I'm used to the docs being at the top and so I missed that when visually scanning, sorry about that.

On a side-note, why doesn't we just use `Option<U64>` with the default serde impls? For convenience, to only be concerned with primitive `u64` after deserializing?

Reply:
I think you're right that the regex pattern is wrong.

A single byte is always a pair of two hexadecimal numbers. Hence `01` or `0f` is a single byte. The first of the two characters in a non-zero byte is not considered to be an individual zero. Only `00` is.

Reply:
We discussed synchronously and we initially agreed that:
1. we will attempt to follow spec (disallow leading zeroes) and only deviate if it brings enough benefit and is supported by Geth (cc @fvictorio)
2. we should do this outside of this PR
3. Geth seems to be parsing hex strings with full bytes, so that's another source of confusion here

Reply:
@Wodann are you sure? This would also accept `0X` prefix as well as octal `0o`/`0O` and binary `0b`/`0B` which I'm pretty sure we do not want to accept, given that we accept to conform only to the spec `0x`: https://github.com/recmo/uint/blob/7b8b0d313a4960091c07eb3356c6a2512a46b585/src/string.rs#L200-L210

Reply:
I don't want to prolong this but since we're already splitting hairs here, it's just worth noting that it'd not be consistent with what we do elsewhere, which is `String::deserialize`, then check for `0x` and then call `U64::from_str` (on a value that's guaranteed to start with `0x`, so okay).

Maybe just being consistent with the rest of the code would be okay, where we do the flow outlined above like everywhere else, and we could focus on compliance/correctness in a separate PR? WDYT?

Reply:
Right, I forgot that we support direct numbers, which is where your suggestion for `U64::deserialize` came from, but we should not accept the different modes, so I tried to make it more explicit what we support and why in the code and in the error message.

Since we also accept JS numbers as the extension to the regular JSON-RPC API, it'd be weird not to support them as a regular string (so also a base 10 number). I've added a comment and made it more explicit about the base, let me know what you think.

Reply:
I only skimmed this discussion (let me know if there's something in particular I should pay attention to) but this caught my eye:

> This would also accept 0X prefix as well as octal 0o/0O and binary 0b/0B which I'm pretty sure we do not want to accept

I checked and the released version of EDR accepts all of these (which is wrong by the spec, and geth doesn't accept):

```
send("eth_getBlockByNumber", [0, false])
send("eth_getBlockByNumber", ["0", false])
send("eth_getBlockByNumber", ["0b0", false])
send("eth_getBlockByNumber", ["0o0", false])
```

I will open an issue about this, but we don't need to do anything about that in this PR.

Reply:
https://github.com/NomicFoundation/edr/issues/509